### ISSUE TITLE
[devstats]The 'Overall Project Statistics Table' of Karmada project not updated in time

### ISSUE THREAD

**Issue by RainbowMango** [cncf/landscape#2976]
Hi, The [ 'Overall Project Statistics Table' metrics](https://karmada.devstats.cncf.io/d/18/overall-project-statistics-table?orgId=1) were probably not updated in time.

I noticed that the `Stargazers` number is always slightly lower than it really is.

| Date | [devstats ](https://karmada.devstats.cncf.io/d/18/overall-project-statistics-table?orgId=1)| [karmada repo](https://github.com/karmada-io/karmada) |
| --- | --- | --- |
| 2023-01-11| 2876| 2984|
| 2023-01-12| 2876| 3003|

The `Stargazers` should be the count overall of all repos in the organization, but from devstats it is even lower than a single repo.

PS:
Could anyone guide me to a link to how these metrics are collected?

**Comment by RobertKielty** [comment]
Hi @lukaszgryglicki can you PTAL and let @RainbowMango how often dashboards are updated? Many thanks.

**Comment by lukaszgryglicki** [comment]
I can take a look tomorrow, if that was reported on `cncf/devstats` I would take a look earlier, because I would get automatic email, here I got email because you tagged me.

**Comment by lukaszgryglicki** [comment]
So, this metric `Stargazers` is coming from counting `WatchEvent`s from GitHub archives.

We only have the following fields there:

- `WatchEvent`: counts the number of distinct actors watched a repo for all repos combined.
- `ForkEvent`: counts number of distinct actors who made a fork for all repos combined.

There is no `Star` even data in GHA so I cannot count stars - this was implemente dthis way since the beginning. See:
```
karmada=# select distinct type from gha_events;
             type              
-------------------------------
 assigned
 closed
 CommitCommentEvent
 connected
 convert_to_draft
 CreateEvent
 DeleteEvent
 demilestoned
 ForkEvent
 head_ref_deleted
 head_ref_force_pushed
 head_ref_restored
 IssueCommentEvent
 IssuesEvent
 labeled
 locked
 marked_as_duplicate
 MemberEvent
 mentioned
 merged
 milestoned
 pinned
 PublicEvent
 PullRequestEvent
 PullRequestReviewCommentEvent
 PullRequestReviewEvent
 PushEvent
 ready_for_review
 referenced
 ReleaseEvent
 renamed
 reopened
 review_requested
 review_request_removed
 subscribed
 unassigned
 unlabeled
 unpinned
 unsubscribed
 WatchEvent
(40 rows)
```

And:
```
karmada=# select distinct type from gha_events where type like '%Event';
             type              
-------------------------------
 CommitCommentEvent
 CreateEvent
 DeleteEvent
 ForkEvent
 IssueCommentEvent
 IssuesEvent
 MemberEvent
 PublicEvent
 PullRequestEvent
 PullRequestReviewCommentEvent
 PullRequestReviewEvent
 PushEvent
 ReleaseEvent
 WatchEvent
(14 rows)
```

Now value for `WatchEvents per repos`:
```
karmada=# select count(distinct actor_id) from gha_events where type = 'WatchEvent';
 count 
-------
  2883
(1 row)

karmada=# select dup_repo_name as repo, count(distinct actor_id) as watches from gha_events where type = 'WatchEvent' group by dup_repo_name order by watches desc;
                  repo                  | watches 
----------------------------------------+---------
 karmada-io/karmada                     |    2866
 karmada-io/dashboard                   |      29
 karmada-io/website                     |      14
 karmada-io/multi-cluster-ingress-nginx |      10
 karmada-io/karmada-operator            |       8
 karmada-io/api                         |       7
 karmada-io/community                   |       6
(7 rows)
```

I'll update the dashboard documentation to include this information.

To sum up: you see 2883 value which is a count of distinct `Watch events` actors across all repos.

**Comment by lukaszgryglicki** [comment]
The [dashboard](https://karmada.devstats.cncf.io/d/18/overall-project-statistics-table?orgId=1) is updated now.
- The etric name is now: `Stargazers/Watchers`.
- Documentation in `Dashboard documentation` now says:
```
Stargazers/Watchers are calculated as sum of distinct actors who performed a watch action on any of repositories that belong to a selected repository group.
```

The confusion was because sum of all distinct watchers for all Karmada org repos was approaching number of stargazers on a single/main repo. Sorry for this confusion and lack of metric description.

**Comment by RainbowMango** [comment]
Thanks @lukaszgryglicki for your quick response. 

> Stargazers/Watchers are calculated as sum of distinct actors who performed a watch action on any of repositories that belong to a selected repository group.

I'm probably even more confused now. From the updated dashboard, I can see the number is `2882`, if it represents the number of users who `watch` the repo, it's a lot more than it actually is.

> There is no Star even data in GHA so I cannot count stars - this was implemente dthis way since the beginning

From the [Stars and Forks by Repository](https://karmada.devstats.cncf.io/d/3/stars-and-forks-by-repository?orgId=1) table, I can see the star stats. I wonder how it was done?

**Comment by lukaszgryglicki** [comment]
- 2888 is the numbe rof distinct actors who ever made a watch to any karmada repo (this is who I see in database and I think it is OK, it comes from events - see queries pasted).
- [this dashboard](https://karmada.devstats.cncf.io/d/3/stars-and-forks-by-repository?orgId=1) uses "fork" events - when anyone does a fork, a "forkee" object is generated and it indeed has numbe rof stargazers at the time of when fork was made, so we can craete that dashboard that gets `max` value for each repo and then sums it, but this is done in that particular dashboard and the one you reported uses count of distinct actors who ever made a watch on any of karmada repos. If you need stargazers then the 2nd dashboard you just linked will give you `approximate number` (3059) - which is a sum of max stargazers from each repo computed from forkee objects.

**Comment by lukaszgryglicki** [comment]
So to sum up - you need to use https://karmada.devstats.cncf.io/d/3/stars-and-forks-by-repository?orgId=1 - the 2nd dashboard you linked for Stargazers - 1st one `cannot` have it becaus eit uses events objects and there is no even to stargazers - we only get stargazers data when fork happens - as an additional column to forkee object, so its resolution is not limited to stargaze events (becaus ethere are none such) but to forks.

**Comment by RainbowMango** [comment]
> 2888 is the numbe rof distinct actors who ever made a watch to any karmada repo

Why should we count and show the people who ever `watched` a repo but not watching the repo now? Frankly speaking, I don't know what I can get from this metric.

In addition, I guess the concept of `Stargazers` should be consistent with GitHub, as well as the [Stars and Forks by Repository](https://karmada.devstats.cncf.io/d/3/stars-and-forks-by-repository?orgId=1) table(2nd dashboard, it seems correct).

**Comment by lukaszgryglicki** [comment]
Well I'm saying how this works right now, we can change this as a feature request.
```
Why should we count and show the people who ever watched a repo but not watching the repo now? 
```
I cannot subtract people who stopped waytching the repo - there is no `unwatch` event - I'm limited to data which I have.
We can even disable this metric in `Karmada` project only if this is a problem for you, but everywhere else this definition is used for years and nobody complained about this. This is the `best afford` that I can get from data.

If you need to `change` something - especially in `Karmada` instance - no problem - please craete a feature request on `cncf/devstats` repo then.

**Comment by RainbowMango** [comment]
Thanks @lukaszgryglicki.

> If you need to change something - especially in Karmada instance

Nope, that's not what I mean. Can you help to revert the change made in the Karmada instance? I think the metrics should be consistent over all projects.

> Well I'm saying how this works right now, we can change this as a feature request.

I totally understand, are you going to put this into the feature plan?

**Comment by lukaszgryglicki** [comment]
I didn't change any metrics, I just updated documentation to specify how it works.
I don't think any feature is needed as the 2nd dashboard you mentioned gives you sum of all stargazers across all repos, you just need to use that 2nd dashboard not this one.
